libraries/map/src/main/kotlin/io/element/android/libraries/map/GeocodingApi.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2023 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.libraries.map
+
+import io.element.android.libraries.map.internal.API_KEY
+import io.element.android.libraries.map.internal.BASE_URL
+import io.element.android.libraries.network.RetrofitFactory
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import retrofit2.create
+import retrofit2.http.GET
+import retrofit2.http.Path
+
+/**
+ * Creates a [GeocodingApi] instance.
+ */
+fun GeocodingApi(
+    retrofitFactory: RetrofitFactory
+): GeocodingApi = retrofitFactory.create(BASE_URL).create() // TODO: Test with mockwebserver?
+
+/**
+ * Creates a fake [GeocodingApi] for tests.
+ */
+fun FakeGeocodingApi(): GeocodingApi = object : GeocodingApi {
+    override suspend fun reverseGeocoding(lat: Double, lon: Double): ReverseGeocodingResponse {
+        return ReverseGeocodingResponse(
+            features = listOf(
+                ReverseGeocodingResponse.Feature(
+                    placeName = "4 Turnham Green Terrace Mews, London Borough of Hounslow, W41QU, United Kingdom"
+                )
+            )
+        )
+    }
+}
+
+/**
+ * https://docs.maptiler.com/cloud/api/geocoding/#search-by-coordinates-reverse
+ */
+interface GeocodingApi {
```

Comment:
Should we separate the retrofit interface from the interface that clients call?

e.g. 

```kotlin
interface Geocoder {
    suspend fun getBestName(lat: Double, lon: Double): String?
}

class GeocoderImpl(retrofitFactory: RetrofitFactory) : Geocoder {
    private val geocodingApi = retrofitFactory.create(BASE_URL).create()
    override suspend fun getBestName(lat: Double, lon: Double): String? {
        geocodingApi.reverseGeocoding(lat, lon).firstPlaceName()
    }
}

class FakeGeocoder : Geocoder { ... }
```

Reply:
Isn't Retrofit all about saving us from doing this (i.e. it creates the impl of the interface at runtime so we don't have to do it, while still allowing us to write fakes for tests)?
Putting another interface in front of the existing one will shield callers from seeing that that we're actually using retrofit, but is this something useful in our context?

Reply:
I was thinking more about shielding the callers from the details of the API, rather than from retrofit itself (e.g. if we added a different backend that used a slightly different response format). Maybe we just tackle that if it happens though :woman_shrugging:

Reply:
Oh I see, definitely, this is something we'd want to tackle in https://github.com/vector-im/element-meta/issues/1722